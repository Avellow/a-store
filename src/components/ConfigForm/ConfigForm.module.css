.configForm {
  display: flex;
  flex-direction: column;
  max-width: 290px;
  gap: 15px;
  padding: 15px 0;
}

.select {
  width: 100%;
}
