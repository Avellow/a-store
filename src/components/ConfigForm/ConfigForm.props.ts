import { DetailedHTMLProps, HTMLAttributes } from "react";
import { CartItemOptionsType, ProductType } from "../../types/api";

type ProductConfigOptionsType = Pick<
  ProductType,
  "sizes" | "colors" | "models" | "stickerNumbers"
>;

type FormProps = DetailedHTMLProps<
  HTMLAttributes<HTMLFormElement>,
  HTMLFormElement
>;

type ConfigFormProps = FormProps & {
  productOptions: ProductConfigOptionsType;
  onConfirm?: (config: CartItemOptionsType) => void;
};

export { ConfigFormProps, ProductConfigOptionsType };
