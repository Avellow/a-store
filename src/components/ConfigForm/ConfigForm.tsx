import { Button } from '@alfalab/core-components/button';
import { BaseSelectChangePayload, OptionShape, Select, Option, OptionProps } from '@alfalab/core-components/select';
import { useEffect, useState } from 'react';
import { ConfigFormProps, SelectOptions } from './ConfigForm.props';
import styles from './ConfigForm.module.css';
import { translateColorRu } from '../../vendor/constants';
import cn from 'classnames';

type IOptions = {
  [P in keyof SelectOptions]: OptionShape[]
}

type SelectedOptions = {
  [P in keyof SelectOptions]: OptionShape
}

type ValueRendererProps = {
  selected?: OptionShape;
  selectedMultiple: OptionShape[];
};

export const ConfigForm = ({ productOptions, className }: ConfigFormProps): JSX.Element => {

  const [options, setOptions] = useState<IOptions>({});
  const [selectedOptions, setSelectedOptions] = useState<SelectedOptions>({});

  useEffect(() => {
    Object
      .entries(productOptions)
      .forEach(([prop, values]) => {
        if (values?.length) {
          const options = values.map((value, i) => ({ key: `${i}`, content: `${value}` }));
          setOptions(prevValue => ({ ...prevValue, [prop]: options }));
          setSelectedOptions(prevValue => ({ ...prevValue, [prop]: options[0] }));
        }
      })
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleOptionChange = ({ name, selected }: BaseSelectChangePayload) => {
    if (name) {
      setSelectedOptions(prevState => ({ ...prevState, [name]: selected }))
    }
  }

  // три функции для перевода названия цвета в Select'е
  const RuColorOption = (props: OptionProps): JSX.Element => (
    <Option {...props}>
      {translateColorRu(props.option.content as string)}
    </Option>
  );

  const SelectedOption = ({ content }: OptionShape) => (
    <span>
      {translateColorRu(content as string)}
    </span>
  )

  const valueRenderer = ({ selected }: ValueRendererProps) =>
    selected ? <SelectedOption {...selected} /> : null

  return (
    <form className={cn(className, styles.configForm)} data-test-id='product-config-form'>
      {
        options.colors && (
          <Select
            className={styles.select}
            onChange={handleOptionChange}
            selected={selectedOptions.colors?.key}
            name='colors'
            label='цвет'
            placeholder='Выберете цвет'
            options={options.colors}
            Option={RuColorOption}
            valueRenderer={valueRenderer}
            dataTestId='color-select'
          />
        )
      }
      {
        options.sizes && (
          <Select
            className={styles.select}
            label='размер'
            onChange={handleOptionChange}
            selected={selectedOptions.sizes?.key}
            name='sizes'
            placeholder='Выберете размер'
            options={options.sizes}
            dataTestId='size-select'
          />
        )
      }
      {
        options.models && (
          <Select
            className={styles.select}
            label='модель'
            selected={selectedOptions.models?.key}
            onChange={handleOptionChange}
            name='models'
            placeholder='Выберете модель'
            options={options.models}
            dataTestId='model-select'
          />
        )
      }
      {
        options.stickerNumbers && (
          <Select
            className={styles.select}
            name='stickerNumbers'
            onChange={handleOptionChange}
            selected={selectedOptions.stickerNumbers?.key}
            label='стикер'
            placeholder='Номер стикера'
            options={options.stickerNumbers}
            dataTestId='sticker-select'
          />
        )
      }
      <Button view='primary' data-test-id='confirm'>
        В корзину
      </Button>
    </form>
  )
}