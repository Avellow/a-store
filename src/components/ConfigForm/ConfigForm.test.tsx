import { render, screen, configure } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { ConfigForm } from './ConfigForm';
import { ConfigFormProps, ProductConfigOptionsType } from './ConfigForm.props';

configure({ testIdAttribute: 'data-test-id' });

const selectableOptions: ProductConfigOptionsType = {
  'colors': ['white', 'black'],
  'sizes': ['M', 'S', 'XXL'],
  'models': undefined,
  'stickerNumbers': [1, 3]
}

const Config = (props: ConfigFormProps): JSX.Element => {
  return (
    <ConfigForm {...props} />
  )
};

describe('Config Form', () => {
  it('should be rendered correctly', () => {
    expect(render(<Config productOptions={selectableOptions} />)).toMatchSnapshot();
  });

  it('should render selects correctly', () => {
    render(<Config productOptions={selectableOptions} />);

    expect(screen.getByTestId('color-select')).toBeInTheDocument();
    expect(screen.getByTestId('size-select')).toBeInTheDocument();
    expect(screen.getByTestId('sticker-select')).toBeInTheDocument();

    expect(screen.queryByTestId('model-select')).not.toBeInTheDocument();
  });

  it('should render button', () => {
    render(<Config productOptions={selectableOptions} />);

    expect(screen.getByTestId('confirm')).toBeInTheDocument();
  });

  it('should render button even if there are no selects', () => {
    render(<Config productOptions={{}} />);

    expect(screen.getByTestId('confirm')).toBeInTheDocument();
  });

  it('should insert the className if it was passed', () => {
    const addClassName = 'additional-class';
    render(<Config productOptions={{}} className={addClassName} />);

    expect(screen.getByTestId('product-config-form')).toHaveClass(addClassName);
  })
});