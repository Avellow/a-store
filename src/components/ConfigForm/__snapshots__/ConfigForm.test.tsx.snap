// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config Form should be rendered correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="configForm"
        data-test-id="product-config-form"
      >
        <div
          aria-haspopup="listbox"
          aria-owns="downshift-1-menu"
          class="select__component_1frn5 select"
          data-test-id="color-select"
          role="combobox"
          tabindex="-1"
        >
          <div
            class="select__component_5l6nj"
          >
            <div
              class="form-control__component_3j482 form-control__component_tvifb form-control__s_3j482 form-control__block_3j482 form-control__hasRightAddons_3j482"
            >
              <div
                aria-controls="downshift-1-menu"
                aria-labelledby="downshift-1-label"
                class="select__field_5l6nj form-control__inner_3j482 form-control__inner_tvifb form-control__filled_3j482 form-control__filled_tvifb form-control__hasInnerLabel_3j482"
                data-test-id="color-select-field"
                id="downshift-0-input"
                tabindex="0"
              >
                <div
                  class="form-control__inputWrapper_3j482"
                >
                  <span
                    aria-hidden="true"
                    class="form-control__hiddenLabel_3j482"
                  >
                    <span
                      for="downshift-0-input"
                      id="downshift-1-label"
                    >
                      цвет
                    </span>
                  </span>
                  <div
                    class="form-control__label_3j482 form-control__label_tvifb"
                  >
                    <span
                      class="form-control__labelInner_3j482"
                    >
                      <span
                        for="downshift-0-input"
                        id="downshift-1-label"
                      >
                        цвет
                      </span>
                    </span>
                  </div>
                  <div
                    class="form-control__input_3j482"
                  >
                    <div
                      class="select__contentWrapper_5l6nj"
                    >
                      <div
                        class="select__value_5l6nj"
                      >
                        <span>
                          белый
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="form-control__addons_3j482 form-control__rightAddons_3j482"
                >
                  <svg
                    class="select__arrow_cajh7 select__arrow_5l6nj"
                    fill="currentColor"
                    focusable="false"
                    height="24"
                    role="img"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <input
            name="colors"
            type="hidden"
            value="0"
          />
        </div>
        <div
          aria-haspopup="listbox"
          aria-owns="downshift-3-menu"
          class="select__component_1frn5 select"
          data-test-id="size-select"
          role="combobox"
          tabindex="-1"
        >
          <div
            class="select__component_5l6nj"
          >
            <div
              class="form-control__component_3j482 form-control__component_tvifb form-control__s_3j482 form-control__block_3j482 form-control__hasRightAddons_3j482"
            >
              <div
                aria-controls="downshift-3-menu"
                aria-labelledby="downshift-3-label"
                class="select__field_5l6nj form-control__inner_3j482 form-control__inner_tvifb form-control__filled_3j482 form-control__filled_tvifb form-control__hasInnerLabel_3j482"
                data-test-id="size-select-field"
                id="downshift-2-input"
                tabindex="0"
              >
                <div
                  class="form-control__inputWrapper_3j482"
                >
                  <span
                    aria-hidden="true"
                    class="form-control__hiddenLabel_3j482"
                  >
                    <span
                      for="downshift-2-input"
                      id="downshift-3-label"
                    >
                      размер
                    </span>
                  </span>
                  <div
                    class="form-control__label_3j482 form-control__label_tvifb"
                  >
                    <span
                      class="form-control__labelInner_3j482"
                    >
                      <span
                        for="downshift-2-input"
                        id="downshift-3-label"
                      >
                        размер
                      </span>
                    </span>
                  </div>
                  <div
                    class="form-control__input_3j482"
                  >
                    <div
                      class="select__contentWrapper_5l6nj"
                    >
                      <div
                        class="select__value_5l6nj"
                      >
                        <span>
                          M
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="form-control__addons_3j482 form-control__rightAddons_3j482"
                >
                  <svg
                    class="select__arrow_cajh7 select__arrow_5l6nj"
                    fill="currentColor"
                    focusable="false"
                    height="24"
                    role="img"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <input
            name="sizes"
            type="hidden"
            value="0"
          />
        </div>
        <div
          aria-haspopup="listbox"
          aria-owns="downshift-5-menu"
          class="select__component_1frn5 select"
          data-test-id="stickerNumber-select"
          role="combobox"
          tabindex="-1"
        >
          <div
            class="select__component_5l6nj"
          >
            <div
              class="form-control__component_3j482 form-control__component_tvifb form-control__s_3j482 form-control__block_3j482 form-control__hasRightAddons_3j482"
            >
              <div
                aria-controls="downshift-5-menu"
                aria-labelledby="downshift-5-label"
                class="select__field_5l6nj form-control__inner_3j482 form-control__inner_tvifb form-control__filled_3j482 form-control__filled_tvifb form-control__hasInnerLabel_3j482"
                data-test-id="stickerNumber-select-field"
                id="downshift-4-input"
                tabindex="0"
              >
                <div
                  class="form-control__inputWrapper_3j482"
                >
                  <span
                    aria-hidden="true"
                    class="form-control__hiddenLabel_3j482"
                  >
                    <span
                      for="downshift-4-input"
                      id="downshift-5-label"
                    >
                      номер стикера
                    </span>
                  </span>
                  <div
                    class="form-control__label_3j482 form-control__label_tvifb"
                  >
                    <span
                      class="form-control__labelInner_3j482"
                    >
                      <span
                        for="downshift-4-input"
                        id="downshift-5-label"
                      >
                        номер стикера
                      </span>
                    </span>
                  </div>
                  <div
                    class="form-control__input_3j482"
                  >
                    <div
                      class="select__contentWrapper_5l6nj"
                    >
                      <div
                        class="select__value_5l6nj"
                      >
                        <span>
                          1
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="form-control__addons_3j482 form-control__rightAddons_3j482"
                >
                  <svg
                    class="select__arrow_cajh7 select__arrow_5l6nj"
                    fill="currentColor"
                    focusable="false"
                    height="24"
                    role="img"
                    viewBox="0 0 24 24"
                    width="24"
                  >
                    <path
                      d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <input
            name="stickerNumbers"
            type="hidden"
            value="0"
          />
        </div>
        <button
          class="button__component_onnkc button__m_onnkc button__component_1qjkj button__primary_1qjkj"
          data-test-id="confirm"
          type="submit"
        >
          <span
            class="button__text_onnkc button__stretchText_onnkc"
          >
            В корзину
          </span>
        </button>
      </form>
    </div>
    <div
      alfa-portal-container=""
    />
  </body>,
  "container": <div>
    <form
      class="configForm"
      data-test-id="product-config-form"
    >
      <div
        aria-haspopup="listbox"
        aria-owns="downshift-1-menu"
        class="select__component_1frn5 select"
        data-test-id="color-select"
        role="combobox"
        tabindex="-1"
      >
        <div
          class="select__component_5l6nj"
        >
          <div
            class="form-control__component_3j482 form-control__component_tvifb form-control__s_3j482 form-control__block_3j482 form-control__hasRightAddons_3j482"
          >
            <div
              aria-controls="downshift-1-menu"
              aria-labelledby="downshift-1-label"
              class="select__field_5l6nj form-control__inner_3j482 form-control__inner_tvifb form-control__filled_3j482 form-control__filled_tvifb form-control__hasInnerLabel_3j482"
              data-test-id="color-select-field"
              id="downshift-0-input"
              tabindex="0"
            >
              <div
                class="form-control__inputWrapper_3j482"
              >
                <span
                  aria-hidden="true"
                  class="form-control__hiddenLabel_3j482"
                >
                  <span
                    for="downshift-0-input"
                    id="downshift-1-label"
                  >
                    цвет
                  </span>
                </span>
                <div
                  class="form-control__label_3j482 form-control__label_tvifb"
                >
                  <span
                    class="form-control__labelInner_3j482"
                  >
                    <span
                      for="downshift-0-input"
                      id="downshift-1-label"
                    >
                      цвет
                    </span>
                  </span>
                </div>
                <div
                  class="form-control__input_3j482"
                >
                  <div
                    class="select__contentWrapper_5l6nj"
                  >
                    <div
                      class="select__value_5l6nj"
                    >
                      <span>
                        белый
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="form-control__addons_3j482 form-control__rightAddons_3j482"
              >
                <svg
                  class="select__arrow_cajh7 select__arrow_5l6nj"
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <input
          name="colors"
          type="hidden"
          value="0"
        />
      </div>
      <div
        aria-haspopup="listbox"
        aria-owns="downshift-3-menu"
        class="select__component_1frn5 select"
        data-test-id="size-select"
        role="combobox"
        tabindex="-1"
      >
        <div
          class="select__component_5l6nj"
        >
          <div
            class="form-control__component_3j482 form-control__component_tvifb form-control__s_3j482 form-control__block_3j482 form-control__hasRightAddons_3j482"
          >
            <div
              aria-controls="downshift-3-menu"
              aria-labelledby="downshift-3-label"
              class="select__field_5l6nj form-control__inner_3j482 form-control__inner_tvifb form-control__filled_3j482 form-control__filled_tvifb form-control__hasInnerLabel_3j482"
              data-test-id="size-select-field"
              id="downshift-2-input"
              tabindex="0"
            >
              <div
                class="form-control__inputWrapper_3j482"
              >
                <span
                  aria-hidden="true"
                  class="form-control__hiddenLabel_3j482"
                >
                  <span
                    for="downshift-2-input"
                    id="downshift-3-label"
                  >
                    размер
                  </span>
                </span>
                <div
                  class="form-control__label_3j482 form-control__label_tvifb"
                >
                  <span
                    class="form-control__labelInner_3j482"
                  >
                    <span
                      for="downshift-2-input"
                      id="downshift-3-label"
                    >
                      размер
                    </span>
                  </span>
                </div>
                <div
                  class="form-control__input_3j482"
                >
                  <div
                    class="select__contentWrapper_5l6nj"
                  >
                    <div
                      class="select__value_5l6nj"
                    >
                      <span>
                        M
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="form-control__addons_3j482 form-control__rightAddons_3j482"
              >
                <svg
                  class="select__arrow_cajh7 select__arrow_5l6nj"
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <input
          name="sizes"
          type="hidden"
          value="0"
        />
      </div>
      <div
        aria-haspopup="listbox"
        aria-owns="downshift-5-menu"
        class="select__component_1frn5 select"
        data-test-id="stickerNumber-select"
        role="combobox"
        tabindex="-1"
      >
        <div
          class="select__component_5l6nj"
        >
          <div
            class="form-control__component_3j482 form-control__component_tvifb form-control__s_3j482 form-control__block_3j482 form-control__hasRightAddons_3j482"
          >
            <div
              aria-controls="downshift-5-menu"
              aria-labelledby="downshift-5-label"
              class="select__field_5l6nj form-control__inner_3j482 form-control__inner_tvifb form-control__filled_3j482 form-control__filled_tvifb form-control__hasInnerLabel_3j482"
              data-test-id="stickerNumber-select-field"
              id="downshift-4-input"
              tabindex="0"
            >
              <div
                class="form-control__inputWrapper_3j482"
              >
                <span
                  aria-hidden="true"
                  class="form-control__hiddenLabel_3j482"
                >
                  <span
                    for="downshift-4-input"
                    id="downshift-5-label"
                  >
                    номер стикера
                  </span>
                </span>
                <div
                  class="form-control__label_3j482 form-control__label_tvifb"
                >
                  <span
                    class="form-control__labelInner_3j482"
                  >
                    <span
                      for="downshift-4-input"
                      id="downshift-5-label"
                    >
                      номер стикера
                    </span>
                  </span>
                </div>
                <div
                  class="form-control__input_3j482"
                >
                  <div
                    class="select__contentWrapper_5l6nj"
                  >
                    <div
                      class="select__value_5l6nj"
                    >
                      <span>
                        1
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="form-control__addons_3j482 form-control__rightAddons_3j482"
              >
                <svg
                  class="select__arrow_cajh7 select__arrow_5l6nj"
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <input
          name="stickerNumbers"
          type="hidden"
          value="0"
        />
      </div>
      <button
        class="button__component_onnkc button__m_onnkc button__component_1qjkj button__primary_1qjkj"
        data-test-id="confirm"
        type="submit"
      >
        <span
          class="button__text_onnkc button__stretchText_onnkc"
        >
          В корзину
        </span>
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
